### ISSUE TITLE
RuntimeException: Trying to call method upper on a non-object

### ISSUE THREAD

**Issue by cottton** [phalcon/cphalcon#14330]
See https://github.com/phalcon/cphalcon/issues/11466#issuecomment-480247019

This is my fault: https://github.com/phalcon/cphalcon/issues/13010

See https://github.com/phalcon/cphalcon/blob/caeac66e7db02fd54c44a27cf22a6381654af2ee/phalcon/db/adapter/pdo.zep#L135

    /**
     * Check for \PDO::XXX class constant aliases
     */
    for key, value in options {
        if typeof key == "string" && defined("\PDO::" . key->upper()) {
            let options[constant("\PDO::" . key->upper())] = value;
            unset options[key];
        }
    }

EDIT: `key->upper()` must be something valid in zep (string in key to upper)

Could somebody pls fix this? 
I do not have enough knowledge in zep - told ya before :p 

**Karma** i just needed this feature and it shot the error msg right in my face :/

**Comment by ruudboon** [comment]
What Phalcon version are you on?

**Comment by cottton** [comment]
3.4
and this feature was `Implemented in the 3.3.x branch.` (https://github.com/phalcon/cphalcon/issues/13010#issuecomment-346689104)

So 3.3, 3.4 and 4. would need that to be fixed.

**Comment by sergeyklay** [comment]
Because `for key, value in options` will create `var`s, not `string`s. In other hand:
```zep
var a = "abc";

typeof a; // string
``` 

But `upper` is present for true `string` datatype, e.g.
```zep
string a = "abc";
a->upper(); // ABC
```

**Comment by cottton** [comment]
So this

        for key, value in options {
            let castValue = (string) key;
            if typeof key == "string" && defined("\PDO::" . castValue->upper()) {
                let options[constant("\PDO::" . castValue->upper())] = value;
                unset options[key];
            }
        }

or this

        for key, value in options {
            if typeof key == "string" && defined("\PDO::" . (string)key->upper()) {
                let options[constant("\PDO::" . (string)key->upper())] = value;
                unset options[key];
            }
        }

?

Still not sure about the rest.
Not going to mess that up twice :D

**Comment by sergeyklay** [comment]
IMO better is to replace by `"\PDO::" . strtoupper(key)`

**Comment by cottton** [comment]
Thanks @ruudboon .
Perhaps a test would be good. 

Dont know how to add but it should check 
the options (`descriptor["options"]`) provided in param `descriptor` :

    // test numeric keys like PDO constants does not get changed
    // f.e. 
    // \PDO::MYSQL_ATTR_INIT_COMMAND => ...
    // which is
    // 1002 => ...
    // must not be changed.
    $descriptor = [
        'options' => [
            \PDO::MYSQL_ATTR_INIT_COMMAND => "SET CHARSET 'utf8';",
        ],
    ];

    // test string keys that match PDO constants does get mapped
    // f.e. 
    // "mysql_attr_init_command" => ...
    // which is
    // \PDO::MYSQL_ATTR_INIT_COMMAND => ...
    // which is
    // 1002 => ...
    // should be mapped.
    $descriptor = [
        'options' => [
            'mysql_attr_init_command' => "SET CHARSET 'utf8';",
        ],
    ];

**Comment by sergeyklay** [comment]
Fixed in the `3.4.x` branch. Feel free to open a new issue if the problem appears again. Thank you for the bug report.

**Comment by GammaGames** [comment]
Is there any ETA on the next release? I see the last release was Jun 30, and (from what I can tell) the most recent changeto  3.4.x was this one. I'm attempting to set the application name for a postgresql connection and this is adding a bit of a roadblock.

**Comment by Jeckerson** [comment]
@GammaGames you can compile from sources `3.4.x` branch and use for your project.
But anyway it is strange that new version wasn't released, since there are some changes in it...

@sergeyklay any answers for questions above?

**Comment by sergeyklay** [comment]
We're not plan to release 3.x in near future. I'm sorry. In other hand, there are simple steps to build Phalcon from source code. This is so easily done that everyone can do it himself. Feel free to ping me  in case of any building problem.

BTW, feature introduced in the https://github.com/phalcon/cphalcon/issues/13010 is optional. DB adapters can operate w/o this feature as usual. Not use it if you don't want compile Phalcon from source code.

**Comment by GammaGames** [comment]
Wait, did you just release a new version? I was attempting to make a docker container that would compile 3.4.x from source!

**Comment by sergeyklay** [comment]
Yeah guys. We've released v3.4.5 right now
